plugins/genesys/fu-genesys-usbhub-firmware.c

```diff
@@ -223,75 +327,90 @@ fu_genesys_usbhub_firmware_parse(FuFirmware *firmware,
 	default:
 		break;
 	}
-
-	/* get static tool string */
 	self->st_static_ts = fu_struct_genesys_ts_static_parse(buf, bufsz, static_ts_offset, error);
 
 	/* deduce code size */
 	switch (self->chip.model) {
 	case ISP_MODEL_HUB_GL3521:
-		code_size = 0x5000;
+		fu_firmware_set_size(firmware, 0x5000);
 		break;
 	case ISP_MODEL_HUB_GL3523: {
 		if (self->chip.revision == 50) {
-			guint8 kbs = 0;
-
-			if (!fu_memread_uint8_safe(buf,
-						   bufsz,
-						   GENESYS_USBHUB_CODE_SIZE_OFFSET,
-						   &kbs,
-						   error))
+			if (!fu_genesys_usbhub_firmware_query_codesize(firmware, buf, bufsz, error))
 				return FALSE;
-			code_size = 1024 * kbs;
 		} else {
-			code_size = 0x6000;
+			fu_firmware_set_size(firmware, 0x6000);
 		}
 		break;
 	}
-	case ISP_MODEL_HUB_GL3590: {
-		guint8 kbs = 0;
-		if (!fu_memread_uint8_safe(buf,
-					   bufsz,
-					   GENESYS_USBHUB_CODE_SIZE_OFFSET,
-					   &kbs,
-					   error))
-			return FALSE;
-		code_size = 1024 * kbs;
-		break;
-	}
+	case ISP_MODEL_HUB_GL3590:
 	case ISP_MODEL_HUB_GL3525: {
-		guint8 kbs = 0;
-		if (!fu_memread_uint8_safe(buf,
-					   bufsz,
-					   GENESYS_USBHUB_CODE_SIZE_OFFSET,
-					   &kbs,
-					   error))
+		if (!fu_genesys_usbhub_firmware_query_codesize(firmware, buf, bufsz, error))
 			return FALSE;
-		code_size = 1024 * kbs;
 		break;
 	}
 	default:
 		break;
 	}
-	fu_firmware_set_size(firmware, code_size);
+	code_size = fu_firmware_get_size(firmware);
 
 	/* calculate checksum */
 	if ((flags & FWUPD_INSTALL_FLAG_IGNORE_CHECKSUM) == 0)
 		if (!fu_genesys_usbhub_firmware_verify(buf, bufsz, code_size, error))
 			return FALSE;
 
 	/* get firmware version */
-	if (!fu_memread_uint16_safe(buf,
-				    bufsz,
-				    GENESYS_USBHUB_VERSION_OFFSET,
-				    &version_raw,
-				    G_LITTLE_ENDIAN,
-				    error))
+	if (!fu_genesys_usbhub_firmware_query_version(firmware, buf, bufsz, error))
 		return FALSE;
-	fu_firmware_set_version_raw(firmware, version_raw);
-	version =
-	    g_strdup_printf("%02x.%02x", (version_raw & 0xFF00U) >> 8, (version_raw & 0x00FFU));
-	fu_firmware_set_version(firmware, version);
+
+	/* parse remaining firmware bytes */
+	if (bufsz - code_size > 0) {
+		FuGenesysFwType fw_type = FU_GENESYS_FW_TYPE_UNKNOWN;
+		gsize fw_offset = offset + code_size;
+		gsize remaining = bufsz - code_size;
+		g_autoptr(FuFirmware) firmware_sub = NULL;
+		g_autoptr(GBytes) blob = NULL;
```

Comment:
i think these two need to move inside the while loop

Reply:
done